### PR TITLE
Added registration view, updated controller and presenter, and the re…

### PR DESCRIPTION
**Pull request by briceallard**
…gister usecase has been changed to accept first and last name of user

### CODE REVIEW

File: lib/app/login/login_controller.dart [hunk 1]
```diff
@@ -41,14 +29,10 @@ class LoginController extends Controller {
...
 
-  void _register() {
-    
-  }
-
   /// Login is successful
   void _loginOnComplete() {
     // TODO: Dismiss any loading events then navigate
-    Navigator.of(_context).pushReplacementNamed('/home');
+    Navigator.of(_context).pushNamed('/home');
```
Comment (ShadyBoukhary) [inline lib/app/login/login_controller.dart]: Why not replacement @briceallard ?

File: lib/app/register/register_controller.dart [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  String _firstName;
+  String _lastName;
+  String _userEmail;
+  String _userPassword;
+  dynamic _context;
+
+  AuthenticationRepository _authenticationRepository;
+  RegisterPresenter _registerPresenter;
+  User _currentUser;
+  User get currentUser => _currentUser;
```
Comment (ShadyBoukhary) [inline lib/app/register/register_controller.dart]: those are not being used

File: lib/app/register/register_view.dart [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+          ),
+        ],
+      ),
+    );
+  }
+
+  void _setAgreedToTOS(bool value) {
+    setState(() {
+      _agreedToTOS = value;
+    });
```
Comment (ShadyBoukhary) [inline lib/app/register/register_view.dart]: These should go in the `RegisterController`

File: lib/data/repositories/data_authentication_repository.dart [hunk 4]
```diff
@@ -30,7 +30,7 @@ class DataAuthenticationRepository implements AuthenticationRepository {
 
   /// Registers a `User` using a [email] and a [password] by making an API call to the server.
   /// It is asynchronous and can throw an `APIException` if the statusCode is not 200.
-  Future<void> register({@required String email, @required String password}) async {
+  Future<void> register({@required String firstName, String lastName, String email, @required String password}) async {
```
Comment (ShadyBoukhary) [inline lib/data/repositories/data_authentication_repository.dart]: Add `@required` to the `lastName` and `email`

File: lib/domain/repositories/authentication_repository.dart [hunk 5]
```diff
@@ -6,7 +6,7 @@ import 'package:hnh/domain/entities/user.dart';
 abstract class AuthenticationRepository {
 
   /// Registers a new user using the provided [username] and [password]
-  Future<void> register({@required String email, @required String password});
+  Future<void> register({@required String firstName, String lastName, String email, @required String password});
```
Comment (ShadyBoukhary) [inline lib/domain/repositories/authentication_repository.dart]: Add `@required` same as above

File: lib/app/register/register_presenter.dart [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+  RegisterPresenter(this._authenticationRepository) {
+    _registerUserCase = RegisterUserCase(_authenticationRepository);
+  }
+
+  void _dispose() {
+    _registerUserCase.dispose();
+  }
+
+  void register({@required String firstName, String lastName, String email, String password}) {
```
Comment (ShadyBoukhary) [inline lib/app/register/register_presenter.dart]: Add `@required` to the rest of the parameters

**Review by ShadyBoukhary — CHANGES_REQUESTED** [review]
Resolve comments

**Comment by ShadyBoukhary** [comment]
[Create Register Presenter](https://trello.com/c/tLsal7eK/29-create-register-presenter)

File: lib/app/login/login_controller.dart (hunk 1 above)
Comment (briceallard) [reply lib/app/login/login_controller.dart]: This is the UI glitch I was telling you about. For some reason, on the create account text beneath this button, the Gesture on tap isn't being recognized until AFTER I click Login, and if I go back a screen afterwards, it works fine. Can't figure it out.

File: lib/app/register/register_controller.dart (hunk 2 above)
Comment (briceallard) [reply lib/app/register/register_controller.dart]: the _registerPresenter is being used to call the OnComplete and OnError mothods? Right?

Comment (briceallard) [reply lib/app/register/register_controller.dart]: The User has been removed.

File: lib/app/login/login_controller.dart (hunk 1 above)
Comment (briceallard) [reply lib/app/login/login_controller.dart]: New Issue created with for bug

File: lib/data/repositories/data_authentication_repository.dart (hunk 4 above)
Comment (briceallard) [reply lib/data/repositories/data_authentication_repository.dart]: Changes made

File: lib/domain/repositories/authentication_repository.dart (hunk 5 above)
Comment (briceallard) [reply lib/domain/repositories/authentication_repository.dart]: Changes made

File: lib/app/register/register_presenter.dart (hunk 6 above)
Comment (briceallard) [reply lib/app/register/register_presenter.dart]: Changes made

File: lib/app/register/register_view.dart (hunk 3 above)
Comment (briceallard) [reply lib/app/register/register_view.dart]: Not sure how to get setState in the controller.

File: lib/app/login/login_controller.dart (hunk 1 above)
Comment (briceallard) [reply lib/app/login/login_controller.dart]: Bug fixed. It was an error with the passed context value.

**Comment by briceallard** [comment]
Changes pushed. Couple issues you mentioned I could not resolve. Take a look when you can!

File: lib/app/register/register_view.dart (hunk 3 above)
Comment (ShadyBoukhary) [reply lib/app/register/register_view.dart]: Use the `callHandler()` that is in the `View`
Pass it the method of the controller like the other examples

Comment (ShadyBoukhary) [reply lib/app/register/register_view.dart]: Every `View` has this method
```
  void callHandler(Function fn) {
    setState(() {
      fn();
    });
  }
```
which allows you to call handlers in the `Controller` and set the state at the same time
@briceallard

**Review by ShadyBoukhary — CHANGES_REQUESTED** [review]
Handler in the view

File: lib/app/register/register_view.dart (hunk 3 above)
Comment (briceallard) [reply lib/app/register/register_view.dart]: Fixed

**Review by ShadyBoukhary — APPROVED** [review]

**Pull request merged**